# Credit Risk Probability of Default - Data Exploration

### Packages and Libraries 

In [1]:
# importing libraries and packages
import pandas as pd
import numpy as np


# visualization libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import time as time 


In [2]:
pd.set_option('display.max_columns', None) # display max number of columns
pd.set_option('display.max_rows',None) # display max number of rows

--------------------------------------------------------------------------------------------------------------------------------

### Loading the dataset

In [3]:
# Loading the dataset in the work environment
dataset_loan = pd.read_csv('./data/raw/Loan_Dataset.csv', low_memory = False)

# viewing a sample rows from the dataset
dataset_loan.sample(5)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
399792  NaN        NaN      27000        27000          27000.0   36 months   
184484  NaN        NaN      40000        40000          40000.0   60 months   
1712993 NaN        NaN      35000        35000          35000.0   36 months   
944239  NaN        NaN      20000        20000          19975.0   36 months   
1517157 NaN        NaN       8000         8000           8000.0   36 months   

         int_rate  installment grade sub_grade                 emp_title  \
399792      11.98       896.53     B        B5          Registered nurse   
184484      12.73       904.61     B        B5              Unit Manager   
1712993     14.30      1201.33     C        C1                  Director   
944239       9.17       637.58     B        B2  Manager of Annual Giving   
1517157     15.05       277.52     C        C4                       DSP   

        emp_length home_ownership  annual_inc verification_status   issue_d  \
399792     2 years           RENT     95000.0            Verified  Mar-2018   
184484   10+ years       MORTGAGE     90500.0            Verified  Aug-2018   
1712993    3 years       MORTGAGE    155000.0     Source Verified  Oct-2013   
944239     4 years       MORTGAGE     78000.0     Source Verified  Dec-2015   
1517157    6 years           RENT     32000.0     Source Verified  Aug-2017   

        loan_status pymnt_plan  url desc             purpose  \
399792      Current          n  NaN  NaN  debt_consolidation   
184484      Current          n  NaN  NaN         credit_card   
1712993  Fully Paid          n  NaN  NaN  debt_consolidation   
944239   Fully Paid          n  NaN  NaN         credit_card   
1517157     Current          n  NaN  NaN  debt_consolidation   

                           title zip_code addr_state    dti  delinq_2yrs  \
399792        Debt consolidation    112xx         NY  21.34          0.0   
184484   Credit card refinancing    245xx         VA  26.65          0.0   
1712993         Debt reconstruct    220xx         VA  20.79          1.0   
944239   Credit card refinancing    601xx         IL  30.38          0.0   
1517157       Debt consolidation    971xx         OR  23.11          0.0   

        earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
399792          Jul-2012             0.0                     NaN   
184484          Sep-2006             1.0                     NaN   
1712993         Feb-1983             0.0                    15.0   
944239          Jul-1999             1.0                     NaN   
1517157         Nov-2008             0.0                     NaN   

         mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
399792                      NaN      18.0      0.0      25392        89.7   
184484                      NaN      11.0      0.0      25801        70.9   
1712993                     NaN      11.0      0.0      33464        80.6   
944239                      NaN      12.0      0.0      39844        65.7   
1517157                     NaN       5.0      0.0       2641        20.3   

         total_acc initial_list_status  out_prncp  out_prncp_inv  \
399792        19.0                   f   19748.44       19748.44   
184484        29.0                   w   37040.83       37040.83   
1712993       31.0                   f       0.00           0.00   
944239        43.0                   w       0.00           0.00   
1517157       11.0                   w    4446.79        4446.79   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
399792    9825.890000          9825.89          7251.56        2574.33   
184484    5399.370000          5399.37          2959.17        2440.20   
1712993  43247.573170         43247.57         35000.00        8247.57   
944239   22958.222009         22929.52         20000.00        2958.22   
1517157   4988.670000          4988.67          3553.21        1435.46   

         total_rec_late_fee  recov

Loan dataset has numerical and categorical features along with few columns with missing values. We will be going through these features one by one to identify their importance and select the most revelvant features for training the model. 

Let's know more about the loan dataset, before proceeding ahead.

In [12]:
# checking the shape of the dataset
print(f"Loan dataset has a shape of {dataset_loan.shape}")

Loan dataset has a shape of (2260668, 145)


So in total we have 145 features with +2 Million observation in the raw dataset which needs to be cleaned and modified according to our observations and requirement for model training and evaluation for a more generalized model which is able to customer default rate in advance before the issuing the loan from the client.  

In [13]:
# basic informaiton about the dataset
dataset_loan.info(memory_usage = 'deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260668 entries, 0 to 2260667
Columns: 145 entries, id to settlement_term
dtypes: float64(105), int64(4), object(36)
memory usage: 5.8 GB


Because of the large number of feature dataframe.info() is not showing the detail breakdown of each feature and it's datatype. 

But we can conculde from the output that that we have 4 features with int64, 105 features with float64 and 36 features that are of object datatypes. 

We will have to check one by one these features for their datatype and convert or make changes accordingly. 

Before we will have to make a fucntion that will help to see the various other information about the no. of observations in the dataset, unique values in the dataset, percentage of missing values and so on, before proceeding ahead. 

**Task**
   - Duplicate values.
   - Correct Datatypes (**To be handle during EDA**)
   - Converting feature into more meaningful format. (**Handle During EDA**)
   - Missing values.

**Check for Duplicated values**

Duplicate rows can cause bias in model training for Porbability of Default(PD), impacting our model performance as they effectively give more/increase weight to certain samples.They may also indicate data collection issues.

Identifying and removing them from the dataset, helps to ensure data integrity.


In [5]:
# check for duplicated values in the loan dataset
dataset_loan.duplicated().sum()

0

As we can see we don't have any duplicate values in the loan dataset.

**Checking for Null Values/ Missing Values in the loan dataset**

Missing values can lead to error during data processing and may cause bias on inaccurate predictions in machine learning models. Detecting them early allows us to decide on appropriate handling methods, such as imputation or removal to maintin data quality.

Dropping rows with complete null values, as it's essential to remove them as we proceed ahead with finding null values in the features column wise.

In [ ]:
# Dropping rows with complete null values.
dataset_loan.dropna(axis = 0, how = 'all')

# checking the shape of the loan dataset after dropping the rows with complete null values.
print(f'Shape of loan dataset after dropping rows with complete null values is {dataset_loan.shape}.')

After comparing the before and after shape of loan dataset we can conclude that there were no rows with complete missing values.

Because of large number of feature in the loan dataset, we will be creating a function that will help to list all the features along with non-null values, null values count, percent of null values and datatype of these features as info() is not able to show all the 145 features.

In [11]:
def dataset_meta_data(dataset):
    
    # dataframe to store the meta information about each features
    dataset_meta = pd.DataFrame(columns = ['Feature_name','Non_Null_Values',
                                           'Null_Values','%_Null_Values',
                                           'Unique_Values','Data_Type'])
    # length of the dataset
    l = len(dataset)
    
    # loop to collect meta data for each feature and append to dataset_meta
    for col in dataset.columns:
        
        # know the null values in the feature
        null_value_count = dataset[col].isnull().sum()
        
        # append the meta data about each feature
        dataset_meta.loc[len(dataset_meta),:] = [col,
                                                 dataset[col].notnull().sum(),
                                                 null_value_count,
                                                 round(null_value_count*100/l,2),
                                                 dataset[col].nunique(),
                                                 dataset[col].dtype
                                                ]
    
    # return the dataset_meta
    return dataset_meta    

In [16]:
# view the meta data of loan dataset using dataset_meta_data()
dataset_meta_desc = dataset_meta_data(dataset_loan)

# print the meta data obtained from the dataset_meta_data() function
dataset_meta_desc

Feature_name Non_Null_Values Null_Values  \
0                                            id               0     1303607   
1                                     member_id               0     1303607   
2                                     loan_amnt         1303607           0   
3                                   funded_amnt         1303607           0   
4                               funded_amnt_inv         1303607           0   
5                                          term         1303607           0   
6                                      int_rate         1303607           0   
7                                   installment         1303607           0   
8                                         grade         1303607           0   
9                                     sub_grade         1303607           0   
10                                    emp_title         1221028       82579   
11                                   emp_length         1228153       75454   
12                               home_ownership         1303607           0   
13                                   annual_inc         1303607           0   
14                          verification_status         1303607           0   
15                                      issue_d         1303607           0   
16                                  loan_status         1303607           0   
17                                   pymnt_plan         1303607           0   
18                                          url               0     1303607   
19                                         desc          123327     1180280   
20                                      purpose         1303607           0   
21                                        title         1288181       15426   
22                                     zip_code         1303606           1   
23                                   addr_state         1303607           0   
24                                          dti         1303295         312   
25                                  delinq_2yrs         1303607           0   
26                             earliest_cr_line         1303607           0   
27                               inq_last_6mths         1303606           1   
28                       mths_since_last_delinq          645783      657824   
29                       mths_since_last_record          221766     1081841   
30                                     open_acc         1303607           0   
31                                      pub_rec         1303607           0   
32                                    revol_bal         1303607           0   
33                                   revol_util         1302797         810   
34                                    total_acc         1303607           0   
35                          initial_list_status         1303607           0   
36                                    out_prncp         1303607           0   
37                                out_prncp_inv         1303607           0   
38                                  total_pymnt         1303607           0   
39                              total_pymnt_inv         1303607           0   
40                              total_rec_prncp         1303607           0   
41                                total_rec_int         1303607           0   
42                           total_rec_late_fee         1303607           0   
43                                   recoveries         1303607           0   
44                      collection_recovery_fee         1303607           0   
45                                 last_pymnt_d         1301347        2260   
46                              last_pymnt_amnt         1303607           0   
47                                 next_pymnt_d               0     1303607   
48                           last_credit_pull_d         1303553          54   
49                   collections_12_mths_ex_med         1303551          56   
50    

As we can see from the above table that there are too many columns with missing values and few of the categorical features with too many categories. We will first have to see deal with the null values in the features and set a threshold for the percent of null values that we can use for the features. 

For setting the threshold for percent of null values in the feature, let's first sort the dataset_meta_desc for better understanding about the null values.

We will view the columns with null values more than 1% as according to dataset_meta_desc we have few features with less than or close to 0%(i.e 100-2000 values missing) and won't make sense to view for this analysis.

In [23]:
# sorting the detaset_meta_desc better understanding of percent of null values 
dataset_meta_desc[dataset_meta_desc['%_Null_Values']>1].sort_values(by = ['%_Null_Values'], ascending = False)

Feature_name Non_Null_Values Null_Values  \
0                                            id               0     2260668   
18                                          url               0     2260668   
1                                     member_id               0     2260668   
134  orig_projected_additional_accrued_interest            8426     2252242   
131                             hardship_length           10613     2250055   
124                             hardship_reason           10613     2250055   
125                             hardship_status           10613     2250055   
126                               deferral_term           10613     2250055   
127                             hardship_amount           10613     2250055   
128                         hardship_start_date           10613     2250055   
129                           hardship_end_date           10613     2250055   
130                     payment_plan_start_date           10613     2250055   
133                        hardship_loan_status           10613     2250055   
132                                hardship_dpd           10613     2250055   
135              hardship_payoff_balance_amount           10613     2250055   
136                hardship_last_payment_amount           10613     2250055   
123                               hardship_type           10613     2250055   
139                   debt_settlement_flag_date           33056     2227612   
140                           settlement_status           33056     2227612   
141                             settlement_date           33056     2227612   
142                           settlement_amount           33056     2227612   
143                       settlement_percentage           33056     2227612   
144                             settlement_term           33056     2227612   
121         sec_app_mths_since_last_major_derog           35942     2224726   
116                          sec_app_revol_util          106184     2154484   
112                    sec_app_earliest_cr_line          108021     2152647   
111                             revol_bal_joint          108020     2152648   
113                      sec_app_inq_last_6mths          108021     2152647   
114                            sec_app_mort_acc          108021     2152647   
115                            sec_app_open_acc          108021     2152647   
117                         sec_app_open_act_il          108021     2152647   
118                       sec_app_num_rev_accts          108021     2152647   
119            sec_app_chargeoff_within_12_mths          108021     2152647   
120          sec_app_collections_12_mths_ex_med          108021     2152647   
55                    verification_status_joint          115730     2144938   
53                             annual_inc_joint          120710     2139958   
54                                    dti_joint          120706     2139962   
19                                         desc          126067     2134601   
29                       mths_since_last_record          359156     1901512   
86                     mths_since_recent_bc_dlq          519701     1740967   
50                  mths_since_last_major_derog          580775     1679893   
88               mths_since_recent_revol_delinq          740359     1520309   
47                                 next_pymnt_d          957061     1303607   
28                       mths_since_last_delinq         1102166     1158502   
65                                      il_util         1191818     1068850   
63                           mths_since_rcnt_il         1350744      909924   
69                                     all_util         1394320      866348   
68                                   max_bal_bc         1394539      866129   
67                                  open_rv_24m         1394539      866129   
71                                       inq_fi         1394539      866129   
72    

We can see that there is a clear distinction in the percent of null values in the loan dataset.

**Observation**

   - **100%** - These are either featues with either unique values or descriptive features like url which won't help in model training.
   
       **Features (`id`, `url`, `member_id`)**
    
   - **98.41-99.53** -> These features are related to depth settlement or hardship features which is out of scope for this model as of now. We can try to make a different model for predicting loan status during depth settlement and hardship type. We can seperate these values from the dataset.
   
       **Features - (`sec_app_mths_since_last_major_derog, debt_settlement_flag_date, settlement_status, settlement_date, settlement_amount, settlement_percentage, settlement_term`)**
       
       
    
   - **94 to 95.22 %** -> These features are mostly talking about the second applicant in the joint application for the loan. Most of the values present will be useful for defining the models for a joint application model. But currently we are making a model for both single and joint application.
   
       **Features - (`desc, annual_inc_joint, dti_joint, verification_status_joint, revol_bal_joint, sec_app_earliest_cr_line, sec_app_inq_last_6mths, sec_app_mort_acc, sec_app_open_acc, sec_app_revol_util, sec_app_open_act_il, sec_app_num_rev_accts, sec_app_chargeoff_within_12_mths, sec_app_collections_12_mths_ex_med`)**
       
    
   - **40.25 to 84.11** -> these features are various information about the customers which is linked to the current loan/ credit provided by the client. As we are making model for predicting the loan default before the issusing the loan we can drop these features.
   
       **Features - (`mths_since_last_delinq, mths_since_last_record, next_pymnt_d, mths_since_last_major_derog, mths_since_rcnt_il, il_util, mths_since_recent_bc_dlq, mths_since_recent_revol_delinq`)**
       
       
    
   - **38.1 %** -> these features talk about the final conditions of the borrowers and have to be evaluated. 
   
       **Features - (`open_acc_6m, open_act_il, open_il_12m, open_il_24m, total_bal_il, open_rv_12m, open_rv_24m, max_bal_bc, all_util, inq_fi, total_cu_tl, inq_last_12m`)**
   
    
   - **1% to 13.07%** -> these features are from various aspect and has to be evaluated one by one and selected according to the importance towards the task at hand.
   
       **Features - (`emp_title, emp_length, title, tot_coll_amt, tot_cur_bal, total_rev_hi_lim, acc_open_past_24mths, avg_cur_bal, bc_open_to_buy, bc_util, mo_sin_old_il_acct, mo_sin_old_rev_tl_op, mo_sin_rcnt_rev_tl_op, mo_sin_rcnt_tl, mort_acc, mths_since_recent_bc, mths_since_recent_inq, num_accts_ever_120_pd, num_actv_bc_tl, num_actv_rev_tl, num_bc_sats, num_bc_tl, num_il_tl, num_op_rev_tl, num_rev_accts, num_rev_tl_bal_gt_0, num_sats, num_tl_120dpd_2m, num_tl_30dpd, num_tl_90g_dpd_24m, num_tl_op_past_12m, pct_tl_nvr_dlq, percent_bc_gt_75, tot_hi_cred_lim, total_bal_ex_mort, total_bc_limit, total_il_high_credit_limit`)**

In [18]:
# knowing the number of columns according to null values segregation. 

# features with 100% null values
null_values_100 = dataset_meta_desc[dataset_meta_desc["%_Null_Values"]==100].iloc[:,0]
print(f'Features with 100% null values are {", ".join(null_values_100)} and count is {len(null_values_100.values)} .')

print("\n")
# features with 98.41 to 99.53% null values 
null_values_98_99 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]>98) & (dataset_meta_desc["%_Null_Values"]<99)].iloc[:,0]
print(f'Features with 98.41 to 99.53% null values are {", ".join(null_values_98_99)} and count is {len(null_values_98_99.values)} .')

print("\n")
# features with 94 to 96% null values 
null_values_94_96 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]>94) & (dataset_meta_desc["%_Null_Values"]<96)].iloc[:,0]
print(f'Features with 94 to 96% null values are {", ".join(null_values_94_96)} and count is {len(null_values_94_96.values)} .')

print("\n")
# features with 40 to 85% null values 
null_values_40_85 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]>40) & (dataset_meta_desc["%_Null_Values"]<85)].iloc[:,0]
print(f'Features with 40 to 85%  null values are {", ".join(null_values_40_85)} and count is {len(null_values_40_85.values)} .')

print("\n")
# features with 38.1% null values 
null_values_38 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]<39) & (dataset_meta_desc["%_Null_Values"]>30)].iloc[:,0]
print(f'Features with 38.1% null values are {", ".join(null_values_38)} and count is {len(null_values_38.values)} .')


print("\n")
# features less than 30% null values 
null_values_30 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]<30) & (dataset_meta_desc["%_Null_Values"]>1)].iloc[:,0]
print(f'Features with 30% null values are {", ".join(null_values_30)} and count is {len(null_values_30.values)} .')



Features with 100% null values are id, member_id, url, next_pymnt_d and count is 4 .


Features with 98.41 to 99.53% null values are annual_inc_joint, dti_joint, verification_status_joint, revol_bal_joint, sec_app_earliest_cr_line, sec_app_inq_last_6mths, sec_app_mort_acc, sec_app_open_acc, sec_app_revol_util, sec_app_open_act_il, sec_app_num_rev_accts, sec_app_chargeoff_within_12_mths, sec_app_collections_12_mths_ex_med and count is 13 .


Features with 94 to 96% null values are  and count is 0 .


Features with 40 to 85%  null values are mths_since_last_delinq, mths_since_last_record, mths_since_last_major_derog, open_acc_6m, open_act_il, open_il_12m, open_il_24m, mths_since_rcnt_il, total_bal_il, il_util, open_rv_12m, open_rv_24m, max_bal_bc, all_util, inq_fi, total_cu_tl, inq_last_12m, mths_since_recent_bc_dlq, mths_since_recent_revol_delinq and count is 19 .


Features with 38.1% null values are  and count is 0 .


Features with 30% null values are emp_title, emp_length, title, to

**Observation**
    
    - Large number of features are having missing values as seen above.
    - Features could be divided according to the bins of percentage of missing values in that particular features.
    
  **[Note]** Most of these features are kpi that are tracked after the issusing of the loan or credit. So it's important to identify these columns and removing them from the dataset before we proceed with the EDA.
    
**Task**

   - Identify the Target Column
   - Identify the feature that are tracked after the issuance of the loan. **[Feature Selection]**
   - Remove these features which are tracked after issusing loan. [**Feature Selection**]
   - Drop the features with too many Null values. [**Feature Selection**]

-------------------------------------------------------------------------------

### Target Variable

We are trying to predict the loan_status variable in the load dataset.

Let's see the value distribution in loan_status variable.

In [8]:
# value distribution of variable loan_status
dataset_loan['loan_status'].value_counts()

Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: loan_status, dtype: int64

According to our business proble, we have to predict the whether the borrower will default before the issuing of the loan or credit line to the customers. So we are only predicting for two categories i.e **Fully Paid and Charged Off** . 

Therefore we will only keep observations with loan_status "Fully Paid" or "Charged Off".

In [9]:
# removing other categories in loan status other than Fully Paid or Charged Off
dataset_loan = dataset_loan.loc[dataset_loan['loan_status'].isin(['Fully Paid','Charged Off'])]

In [32]:
# checking the value distribution between "Fully Paid" and "Charged Off"
dataset_loan['loan_status'].value_counts(dropna = False)

Fully Paid     1041952
Charged Off     261655
Name: loan_status, dtype: int64

In [31]:
# checking the normalized distribution between Fully Paid and Charged Pff
dataset_loan['loan_status'].value_counts(normalize = True, dropna = True)

Fully Paid     0.799284
Charged Off    0.200716
Name: loan_status, dtype: float64

**Observations**
 - Close to 80% of the loans are Fully Paid and around 20% of the loans are Charged Off.
 - We will have to handle the unbalance dataset that we have before training either using cross validation or SMOTE method for optimal performance for our model.
 
**Task**
  - Handle imbalance between the two categories using 
      - Cross Validation method. 
      - Downsampling or upsampling using SMOTE method.

-------------------------------------------------------------------------------

We have done the initial checks for our loan dataset and according to the observations we have few of the tasks that needs to be performed as we move forward. 

We will move with Feature Selection and EDA in the below section

## Data Cleaning - Initial Feature Dropping

Raw loan dataset has 145 features and as per out pervious analysis according to the business problem we have not all the features are useful. 

Definition of the features is provided in the excel file here in data/raw/DataDictionary.xlsx .

We will be dropping features with the following criteria. 
   -  Feature with more than 50% of the data missing.
   -  Feature that won't be available at the time of loan application (i.e credit repayement behaviour features.).

### **Features with too much data missing**

In [17]:
# view the features in descending order to the % of null values
dataset_meta_desc.sort_values(by = '%_Null_Values', ascending = False)

Feature_name Non_Null_Values Null_Values  \
0                                            id               0     1303607   
47                                 next_pymnt_d               0     1303607   
1                                     member_id               0     1303607   
18                                          url               0     1303607   
134  orig_projected_additional_accrued_interest            3433     1300174   
124                             hardship_reason            5335     1298272   
130                     payment_plan_start_date            5335     1298272   
123                               hardship_type            5335     1298272   
125                             hardship_status            5335     1298272   
127                             hardship_amount            5335     1298272   
128                         hardship_start_date            5335     1298272   
129                           hardship_end_date            5335     1298272   
126                               deferral_term            5335     1298272   
131                             hardship_length            5335     1298272   
136                hardship_last_payment_amount            5335     1298272   
133                        hardship_loan_status            5335     1298272   
135              hardship_payoff_balance_amount            5335     1298272   
132                                hardship_dpd            5335     1298272   
121         sec_app_mths_since_last_major_derog            5932     1297675   
116                          sec_app_revol_util           16218     1287389   
111                             revol_bal_joint           16510     1287097   
112                    sec_app_earliest_cr_line           16511     1287096   
120          sec_app_collections_12_mths_ex_med           16511     1287096   
119            sec_app_chargeoff_within_12_mths           16511     1287096   
118                       sec_app_num_rev_accts           16511     1287096   
117                         sec_app_open_act_il           16511     1287096   
115                            sec_app_open_acc           16511     1287096   
114                            sec_app_mort_acc           16511     1287096   
113                      sec_app_inq_last_6mths           16511     1287096   
55                    verification_status_joint           23091     1280516   
54                                    dti_joint           23235     1280372   
53                             annual_inc_joint           23237     1280370   
140                           settlement_status           31989     1271618   
143                       settlement_percentage           31989     1271618   
142                           settlement_amount           31989     1271618   
139                   debt_settlement_flag_date           31989     1271618   
141                             settlement_date           31989     1271618   
144                             settlement_term           31989     1271618   
19                                         desc          123327     1180280   
29                       mths_since_last_record          221766     1081841   
86                     mths_since_recent_bc_dlq          309146      994461   
50                  mths_since_last_major_derog          342367      961240   
65                                      il_util          432668      870939   
88               mths_since_recent_revol_delinq          435835      867772   
63                           mths_since_rcnt_il          486548      817059   
72                                  total_cu_tl          499544      804063   
71                                       inq_fi          499545      804062   
64                                 total_bal_il          499545      804062   
62                                  open_il_24m          499545      804062   
61                                  open_il_12m          499545      804062   
60    

In [30]:
# count of features with null values
print(f"Lona Dataset has {dataset_meta_desc.shape[0]} and out of which {dataset_meta_desc[dataset_meta_desc['%_Null_Values']>0].shape[0]} have null values")

Lona Dataset has 145 and out of which 99 have null values


 After the removal of other categories in loan_status feature, when we check the %_Null_Values in the load dataset we can see **clear distinction** between feature with **50%** missing values comapared to **13%** missing values. 
 
 Hence we decide to make a **threshold of 50%** for dropping the features from further process.

In [53]:
# selecting feature with more than 50% of null values
null_feature_50_list = dataset_meta_desc.loc[dataset_meta_desc['%_Null_Values']>50, 'Feature_name'].to_list()

# view the feature names with more than 50% missing values
print(f"Following is the list of all the feature with missing values more than 50% \n\n {', '.join(null_feature_50_list)}")

print("\n")
# view the count of the these features with 50% missing values 
print(f"Total number of features with 50% missing values is {len(null_feature_50_list)}")

Following is the list of all the feature with missing values more than 50% 

 id, member_id, url, desc, mths_since_last_delinq, mths_since_last_record, next_pymnt_d, mths_since_last_major_derog, annual_inc_joint, dti_joint, verification_status_joint, open_acc_6m, open_act_il, open_il_12m, open_il_24m, mths_since_rcnt_il, total_bal_il, il_util, open_rv_12m, open_rv_24m, max_bal_bc, all_util, inq_fi, total_cu_tl, inq_last_12m, mths_since_recent_bc_dlq, mths_since_recent_revol_delinq, revol_bal_joint, sec_app_earliest_cr_line, sec_app_inq_last_6mths, sec_app_mort_acc, sec_app_open_acc, sec_app_revol_util, sec_app_open_act_il, sec_app_num_rev_accts, sec_app_chargeoff_within_12_mths, sec_app_collections_12_mths_ex_med, sec_app_mths_since_last_major_derog, hardship_type, hardship_reason, hardship_status, deferral_term, hardship_amount, hardship_start_date, hardship_end_date, payment_plan_start_date, hardship_length, hardship_dpd, hardship_loan_status, orig_projected_additional_accrued_intere

Above we can see that there are 58 feature with 50% missing values with their name. We will be dropping these features from the load dataset.

**Dropping Features with 50% missing values.**

In [57]:
# dropping the features with more than 50% missing values.
dataset_loan.drop(columns = null_feature_50_list, axis = 1, inplace = True)

# Shape of the loat dataset after dropping above features 
dataset_loan.shape

(1303607, 87)

After dropping the above mentioned features we are left with 87 features in the load dataset.

### Feature not available during loan application.

In our current form of the loan dataset we have features that are being used to track the latest update on the loan/ credit line of the borrowers which won't be present at the moment of loan application.

Hence we can drop these features from our loan dataset. 

In [61]:
# view all the feature present in the loan dataset
print(f" Following is the list of all the features present in loan dataset \n {dataset_loan.columns}")

 Following is the list of all the features present in loan dataset 
 Index(['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'pymnt_plan', 'purpose', 'title', 'zip_code',
       'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util',
       'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_d', 'last_pymnt_amnt', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'policy_code', 'application_type',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',


Upon checking the definition of these features, we will drop the features that would not been available before lending of the loan. 

In [65]:
# list of the feature that were not present before lending of loan.
drop_feature_list = ['acc_now_delinq', 'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'collection_recovery_fee', 'collections_12_mths_ex_med', 'debt_settlement_flag', 'delinq_2yrs', 'delinq_amnt', 'disbursement_method', 'funded_amnt', 'funded_amnt_inv', 'hardship_flag', 'inq_last_6mths', 'last_credit_pull_d', 'last_pymnt_amnt', 'last_pymnt_d', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mths_since_recent_bc', 'mths_since_recent_inq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 'num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m', 'num_tl_30dpd', 'num_tl_90g_dpd_24m', 'num_tl_op_past_12m',  'out_prncp', 'out_prncp_inv', 'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pymnt_plan', 'recoveries', 'tax_liens', 'tot_coll_amt', 'tot_cur_bal', 'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit', 'total_il_high_credit_limit', 'total_pymnt', 'total_pymnt_inv', 'total_rec_int', 'total_rec_late_fee', 'total_rec_prncp', 'total_rev_hi_lim']

# drop features not present before lending of loan. 
dataset_loan.drop(columns = drop_feature_list, axis = 1, inplace = True)

Let's view the number of features present now and their names.

In [4]:
# view the features names present in the loat dataset.
print(f'List of the features present in the loan dataset is {len(dataset_loan.columns)} \n\n {", ".join(sorted(dataset_loan.columns))} ')

List of the features present in the loan dataset is 31 

 addr_state, annual_inc, application_type, dti, earliest_cr_line, emp_length, emp_title, grade, home_ownership, initial_list_status, installment, int_rate, issue_d, loan_amnt, loan_status, mo_sin_old_il_acct, mo_sin_old_rev_tl_op, mort_acc, open_acc, policy_code, pub_rec, pub_rec_bankruptcies, purpose, revol_bal, revol_util, sub_grade, term, title, total_acc, verification_status, zip_code 


## Exploratory Data Analysis

We will start with exploring the loan dataset after doing basic cleaning of removing features with too many null values and dropping the features that won't be available during the application of loan or credit line.

Before we start with EDA, let's view a sample dataset.

In [9]:
# view sample dataset from the loan dataset
dataset_loan.sample(5)


loan_amnt        term  int_rate  installment grade sub_grade  \
659439       8750   36 months     10.49       284.36     B        B3   
913371      20000   36 months      8.90       635.07     A        A5   
286179       2100   36 months     18.99        76.97     D        D3   
894919      25475   36 months     13.68       866.73     C        C1   
29229       30000   36 months     12.61      1005.20     C        C1   

                             emp_title emp_length home_ownership  annual_inc  \
659439                   peace officer  10+ years           RENT     53000.0   
913371  New York Presbyterian Hospital  10+ years       MORTGAGE     62000.0   
286179                            PA-C  10+ years       MORTGAGE    130000.0   
894919                       HomeTelos    6 years       MORTGAGE     65000.0   
29229                Massage Therapist    4 years           RENT     54310.0   

       verification_status   issue_d  loan_status         purpose  \
659439     Source Verified  Feb-2015   Fully Paid     credit_card   
913371            Verified  Jun-2013   Fully Paid     credit_card   
286179            Verified  Nov-2016   Fully Paid           other   
894919            Verified  Jul-2013  Charged Off           other   
29229      Source Verified  Mar-2018  Charged Off  major_purchase   

                          title zip_code addr_state    dti earliest_cr_line  \
659439  Credit card refinancing    100xx         NY  12.88         Aug-2002   
913371                  Control    104xx         NY  11.26         May-1991   
286179                    Other    903xx         CA  22.23         Sep-1995   
894919                 Personal    190xx         PA  11.50         Dec-1997   
29229            Major purchase    282xx         NC   0.69         Aug-2006   

        open_acc  pub_rec  revol_bal  revol_util  total_acc  \
659439       7.0      0.0       8021        36.0       15.0   
913371       6.0      0.0      10124        55.3       23.0   
286179      10.0      0.0      14323        80.8       17.0   
894919      13.0      0.0       3477         5.8       30.0   
29229       12.0      0.0          0         0.0       31.0   

       initial_list_status  policy_code application_type  mo_sin_old_il_acct  \
659439                   f            1       Individual                86.0   
913371                   f            1       Individual                73.0   
286179                   f            1       Individual               254.0   
894919                   f            1       Individual               130.0   
29229                    w            1       Individual               139.0   

        mo_sin_old_rev_tl_op  mort_acc  pub_rec_bankruptcies  
659439                 149.0       0.0                   0.0  
913371                 268.0       3.0                   0.0  
286179                 137.0       3.0                   0.0  
894919                 189.0       3.0                   0.0  
29229                  130.0       0.0                   0.0

In [10]:
# view the shape of the dataset
print(f'Loan dataset has a following number of observations and features {dataset_loan.shape}')

Loan dataset has a following number of observations and features (1303607, 31)


After the initial cleaning of the dataset we are left with 130K observations and 31 features which we will analyze one by one, try to check for datatypes, transformations needed for the features, check for outliers, distribution of the dataset and perform bivariate analysis.

**Checking for Null Values**

In [15]:
# view the meta data of loan dataset using dataset_meta_data()
dataset_meta_desc_1 = dataset_meta_data(dataset_loan)

# print the meta data obtained from the dataset_meta_data() function
dataset_meta_desc_1

Feature_name Non_Null_Values Null_Values %_Null_Values  \
0              loan_amnt         1303607           0           0.0   
1                   term         1303607           0           0.0   
2               int_rate         1303607           0           0.0   
3            installment         1303607           0           0.0   
4                  grade         1303607           0           0.0   
5              sub_grade         1303607           0           0.0   
6              emp_title         1221028       82579          6.33   
7             emp_length         1228153       75454          5.79   
8         home_ownership         1303607           0           0.0   
9             annual_inc         1303607           0           0.0   
10   verification_status         1303607           0           0.0   
11               issue_d         1303607           0           0.0   
12           loan_status         1303607           0           0.0   
13               purpose         1303607           0           0.0   
14                 title         1288181       15426          1.18   
15              zip_code         1303606           1           0.0   
16            addr_state         1303607           0           0.0   
17                   dti         1303295         312          0.02   
18      earliest_cr_line         1303607           0           0.0   
19              open_acc         1303607           0           0.0   
20               pub_rec         1303607           0           0.0   
21             revol_bal         1303607           0           0.0   
22            revol_util         1302797         810          0.06   
23             total_acc         1303607           0           0.0   
24   initial_list_status         1303607           0           0.0   
25           policy_code         1303607           0           0.0   
26      application_type         1303607           0           0.0   
27    mo_sin_old_il_acct         1199312      104295           8.0   
28  mo_sin_old_rev_tl_op         1236079       67528          5.18   
29              mort_acc         1256326       47281          3.63   
30  pub_rec_bankruptcies         1302910         697          0.05   

   Unique_Values Data_Type  
0           1553     int64  
1              2    object  
2            654   float64  
3          82574   float64  
4              7    object  
5             35    object  
6         371435    object  
7             11    object  
8              6    object  
9          62988   float64  
10             3    object  
11           139    object  
12             2    object  
13            14    object  
14         61678    object  
15           943    object  
16            51    object  
17          6870   float64  
18           738    object  
19            84   float64  
20            37   float64  
21         82819     int64  
22          1363   float64  
23           142   float64  
24             2    object  
25             1     int64  
26             2    object  
27           520   float64  
28           754   float64  
29            38   float64  
30            12   float64

In [18]:
# viewing the meta data about the features in loan dataset in descending order
dataset_meta_desc_1.loc[dataset_meta_desc_1['Null_Values']>0].sort_values(by = '%_Null_Values', ascending = False)

Feature_name Non_Null_Values Null_Values %_Null_Values  \
27    mo_sin_old_il_acct         1199312      104295           8.0   
6              emp_title         1221028       82579          6.33   
7             emp_length         1228153       75454          5.79   
28  mo_sin_old_rev_tl_op         1236079       67528          5.18   
29              mort_acc         1256326       47281          3.63   
14                 title         1288181       15426          1.18   
22            revol_util         1302797         810          0.06   
30  pub_rec_bankruptcies         1302910         697          0.05   
17                   dti         1303295         312          0.02   
15              zip_code         1303606           1           0.0   

   Unique_Values Data_Type  
27           520   float64  
6         371435    object  
7             11    object  
28           754   float64  
29            38   float64  
14         61678    object  
22          1363   float64  
30            12   float64  
17          6870   float64  
15           943    object

### Univariate Analysis

We will be starting with univariate analysis of each features one by one and check for following parameters before moving to bivariate analysis and note down any task or featuring engineering task that needs to be performed. 
 - Check for distriubution of each features
 - check for outliers
 - Check for missing values and decided method to remove them according to the presence of outliers in the feature and whether it's numerical or categorical features.

In [20]:
# get the features that have numeric dtypes
numeric_features = [col for col in dataset_loan.columns if dataset_loan[col].dtypes != 'object']
# view the name of these Numeric and Categorical features 
print(f'Following are the numerical columns {numeric_features}')

# get the features with object dtypes
categorical_features = dataset_loan.select_dtypes('object').columns.to_list()
print(f'\nFollowing are the categorical columns {categorical_features}')

Following are the numerical columns ['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'policy_code', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mort_acc', 'pub_rec_bankruptcies']

Following are the categorical columns ['term', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'verification_status', 'issue_d', 'loan_status', 'purpose', 'title', 'zip_code', 'addr_state', 'earliest_cr_line', 'initial_list_status', 'application_type']


**Basic Statistic of the loan dataset**

Let's view the basic statistics of the numerical and categorical features in the loan dataset we have.

In [23]:
# view the basic statistics  of the numerical features.
dataset_loan.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  1.303607e+06  1.303607e+06  1.303607e+06  1.303607e+06  1.303295e+06   
mean   1.441681e+04  1.325722e+01  4.380850e+02  7.615899e+04  1.825875e+01   
std    8.699542e+03  4.760544e+00  2.610632e+02  7.004890e+04  1.093849e+01   
min    5.000000e+02  5.310000e+00  4.930000e+00  0.000000e+00 -1.000000e+00   
25%    8.000000e+03  9.750000e+00  2.488200e+02  4.584000e+04  1.179000e+01   
50%    1.200000e+04  1.274000e+01  3.754300e+02  6.500000e+04  1.761000e+01   
75%    2.000000e+04  1.599000e+01  5.804500e+02  9.000000e+04  2.404000e+01   
max    4.000000e+04  3.099000e+01  1.719830e+03  1.099920e+07  9.990000e+02   

           open_acc       pub_rec     revol_bal    revol_util     total_acc  \
count  1.303607e+06  1.303607e+06  1.303607e+06  1.302797e+06  1.303607e+06   
mean   1.159234e+01  2.155220e-01  1.623599e+04  5.191014e+01  2.501894e+01   
std    5.465282e+00  6.028619e-01  2.228629e+04  2.450155e+01  1.199584e+01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  2.000000e+00   
25%    8.000000e+00  0.000000e+00  5.952000e+03  3.360000e+01  1.600000e+01   
50%    1.100000e+01  0.000000e+00  1.114000e+04  5.230000e+01  2.300000e+01   
75%    1.400000e+01  0.000000e+00  1.974800e+04  7.080000e+01  3.200000e+01   
max    9.000000e+01  8.600000e+01  2.904836e+06  8.923000e+02  1.760000e+02   

       policy_code  mo_sin_old_il_acct  mo_sin_old_rev_tl_op      mort_acc  \
count    1303607.0        1.199312e+06          1.236079e+06  1.256326e+06   
mean           1.0        1.258357e+02          1.814992e+02  1.676875e+00   
std            0.0        5.217895e+01          9.433621e+01  2.005056e+00   
min            1.0        0.000000e+00          2.000000e+00  0.000000e+00   
25%            1.0        9.800000e+01          1.170000e+02  0.000000e+00   
50%            1.0        1.290000e+02          1.640000e+02  1.000000e+00   
75%            1.0        1.520000e+02          2.300000e+02  3.000000e+00   
max            1.0        9.990000e+02          8.520000e+02  5.100000e+01   

       pub_rec_bankruptcies  
count          1.302910e+06  
mean           1.346985e-01  
std            3.782756e-01  
min            0.000000e+00  
25%            0.000000e+00  
50%            0.000000e+00  
75%            0.000000e+00  
max            1.200000e+01

In [24]:
# viewing the basic statistics of the categorical features.
dataset_loan[categorical_features].describe()

term    grade sub_grade emp_title emp_length home_ownership  \
count      1303607  1303607   1303607   1221028    1228153        1303607   
unique           2        7        35    371435         11              6   
top      36 months        B        C1   Teacher  10+ years       MORTGAGE   
freq        988754   380152     82815     20496     428547         645496   

       verification_status   issue_d loan_status             purpose  \
count              1303607   1303607     1303607             1303607   
unique                   3       139           2                  14   
top        Source Verified  Oct-2015  Fully Paid  debt_consolidation   
freq                503726     42727     1041952              757591   

                     title zip_code addr_state earliest_cr_line  \
count              1288181  1303606    1303607          1303607   
unique               61678      943         51              738   
top     Debt consolidation    945xx         CA         Aug-2001   
freq                638945    14625     191032             9153   

       initial_list_status application_type  
count              1303607          1303607  
unique                   2                2  
top                      w       Individual  
freq                751214          1280370

It's necessary to view unique values and their count in each categorical features. 

In [37]:
# view unique and their count in each categorical features.
for col in dataset_loan.select_dtypes('object').columns:
    print(col.upper() , ' : ', dataset_loan[col].nunique())
    if dataset_loan[col].nunique()>51:
        print(dataset_loan[col].unique()[:51])
        print("----- and", len(dataset_loan[col].unique())-51, "more",'\n')
    else:
        print(dataset_loan[col].unique(),'\n')

TERM  :  2
[' 36 months' ' 60 months'] 

GRADE  :  7
['D' 'C' 'A' 'B' 'E' 'G' 'F'] 

SUB_GRADE  :  35
['D5' 'C4' 'A3' 'B3' 'E5' 'D1' 'C1' 'A4' 'E1' 'B5' 'E4' 'B1' 'D2' 'B4'
 'B2' 'C3' 'A5' 'C2' 'A2' 'C5' 'D3' 'D4' 'E2' 'A1' 'E3' 'G2' 'F1' 'G4'
 'F2' 'F4' 'F3' 'F5' 'G1' 'G5' 'G3'] 

EMP_TITLE  :  371435
['Supervisor ' 'Assistant to the Treasurer (Payroll)' 'Teacher'
 'Accounts Examiner III' 'Senior Director Risk Management'
 'Front office Lead' 'Sewell collision center' 'Manager' 'service advisor'
 'Stylist' 'HR Director ' 'Cabinet maker' 'Associate Research Fellow'
 'Supply chain management ' 'Office Assistant 1' 'Deputy'
 'kitchen and bath designer' 'Executive Casino Host' 'Customer service '
 'District Manager' 'Director of Maintenance ' 'Teacher and Coach'
 'Table Games Dealer ' nan 'QUALITY ASSURANCE' 'Mobile Expert'
 'Customer Advocate' 'Verizon Engineer' 'Dietary Supervisor ' 'operator '
 'Claims manager ' 'Electric Systems Associate ' 'Civil Engineering Tech'
 'Systems Administr

**Tasks**
   - adawdawd 
   - 
   - 
  
  
  


### Loading Intermediate data

In [3]:
dataset_loan = pd.read_csv('./data/processed/temp_processed_data_1.csv')

### Saving intermediate data

In [76]:
dataset_loan.to_csv('./data/processed/temp_processed_data_1.csv', index = False)